File: src/lib/stores/project-cache.ts
/**
 * Project snapshot cache — loads/saves per-project state for instant UI.
 *
 * On project switch: load snapshot → apply to titlebar/badges.
 * After real data loads: build snapshot from current state → save to cache.
 *
 * ── Migrated to the RepoState container (spec 08 step 5) ──────────────
 * The in-memory mirror used to be a central `Map<projectPath, ProjectSnapshot>`
 * (`projectSnapshots`) owned here — the exact "wrote under the wrong key"
 * hazard this module documents. It now lives per-repo in `RepoState.snapshot`,
 * reached by path through the container (`getRepoState` / `repoField`), so a
 * RepoState only ever knows its own path. The disk snapshot persistence
 * (`get_project_snapshot` / `save_project_snapshot`) stays here unchanged.
 */

import type { ProjectSnapshot, GraphViewport } from "$lib/types";
import * as api from "$lib/api/tauri";
import { fileStatuses } from "./changes";
import { get, type Readable } from "svelte/store";
import { repoInfo } from "./repo";
import { viewport, graphOffset } from "./graph";
import { getRepoState, repoField } from "./repo-state";

/**
 * Persistent graph viewport slice.
 *
 * Stored inside `ProjectSnapshot.graph_viewport_cache` so a cold start
 * can paint the commit graph synchronously from disk. Size is roughly
 * 60 KB per project (300 rows × ~200 B JSON each).
 */
export interface GraphViewportCache {
  /** Last-seen 300-row viewport window (the `nodes` array as served by
   *  `get_graph_viewport`). No lane segments / merge curves — the fresh
   *  refresh repopulates those within one tick of paint. */
  nodes: GraphViewport["nodes"];
  /** Total commit count for the repo — used to render the scroll footer. */
  total_count: number;
  /** HEAD OID at the time the cache was written. Used as a coarse
   *  staleness check alongside `top_oid`. */
  head_oid: string;
  /** First visible commit in the cached window. Primary staleness
   *  check during reconciliation: if fresh `top_oid` matches, the
   *  cache is still accurate and no repaint is needed. */
  top_oid: string;
  /** Scroll offset captured at cache time — preserves vertical scroll
   *  position across cold starts. */
  offset: number;
  /** Epoch milliseconds when the cache was written. Entries older than
   *  `GRAPH_CACHE_TTL_MS` at load time are ignored and overwritten. */
  cached_at: number;
}

/** Max age before a cached graph slice is ignored at load time (7 days). */
export const GRAPH_CACHE_TTL_MS = 7 * 24 * 60 * 60 * 1000;

/**
 * Return `true` when the cache timestamp is within the TTL window.
 * Boundary (`cached_at === now - TTL`) is accepted so borderline entries
 * hydrate instead of being discarded on a refresh.
 */
export function isCacheFresh(cachedAt: number): boolean {
  return Date.now() - cachedAt <= GRAPH_CACHE_TTL_MS;
}

/**
 * Reactive view of a specific repo's in-memory snapshot, resolved by path
 * through the container (`RepoState.snapshot`). Emits `null` when the repo has
 * no live state (closed tab) or hasn't been hydrated yet. The tab strip
 * subscribes to this per tab so an inactive tab's badges update reactively when
 * a watcher-driven `saveCurrentSnapshot` rewrites that repo's snapshot.
 */
export function snapshotStore(path: string): Readable<ProjectSnapshot | null> {
  return repoField(path, (rs) => rs.snapshot);
}

/**
 * Mirror a snapshot into the owning repo's slice. No-ops when the repo has no
 * live state (a closed/never-opened tab) — the disk read still returns the
 * value to the caller.
 */
export function hydrateSnapshotCache(path: string, snap: ProjectSnapshot): void {
  getRepoState(path)?.snapshot.set(snap);
}

/**
 * Look up a cached snapshot synchronously from the owning repo's slice.
 * Returns `null` when the project has no live state or has never been loaded in
 * this session; callers that need disk I/O should use `loadProjectSnapshot`.
 */
export function getCachedSnapshot(path: string): ProjectSnapshot | null {
  const rs = getRepoState(path);
  return rs ? get(rs.snapshot) : null;
}

/** Load a snapshot from cache. Returns null if not cached. */
export async function loadProjectSnapshot(path: string): Promise<ProjectSnapshot | null> {
  try {
    const snap = await api.getProjectSnapshot(path);
    if (snap) hydrateSnapshotCache(path, snap);
    return snap;
  } catch {
    return null;
  }
}

/**
 * Hydrate `viewport` + `graphOffset` synchronously from the in-memory
 * snapshot cache when a fresh-enough slice exists. Returns `true` when
 * a paint-worthy viewport was installed — i.e. the caller must NOT
 * clobber it with a spinner/skeleton.
 *
 * Callers must have primed the cache (via `loadProjectSnapshot` on a
 * prior activation or an explicit warm-up) — this function never
 * touches disk. The tab-switch path in `projects.ts` first checks the
 * incoming repo's in-memory `GraphSlice` viewport; only when that is
 * empty (cold start) does it fall through here to the disk-backed slice.
 *
 * Writes go to the target repo's own `GraphSlice` (resolved by path), not the
 * active facade — so this can only ever install a viewport under its own key.
 *
 * The restored viewport has empty `lane_segments` / `merge_curves`
 * because the cache doesn't persist layout geometry — the skeleton +
 * node list still renders usefully until the fresh refresh arrives
 * (< 100 ms) and the reconciler swaps in the full geometry.
 */
export function restorePersistedViewport(projectPath: string): boolean {
  const rs = getRepoState(projectPath);
  const snap = rs ? get(rs.snapshot) : null;
  if (!rs || !snap?.graph_viewport_cache) return false;
  const cache = snap.graph_viewport_cache;
  if (!isCacheFresh(cache.cached_at)) return false;
  rs.graph.viewport.set({
    nodes: cache.nodes,
    lane_segments: [],
    merge_curves: [],
    total_count: cache.total_count,
    offset: cache.offset,
    visible_lane_count: 0,
    total_lane_count: 0,
    head_lane: null,
    has_more: false,
  });
  rs.graph.graphOffset.set(cache.offset);
  return true;
}

/**
 * Get a snapshot for any project. Reads the persisted on-disk cache
 * for `path` (fast path); never falls back to live status fetched
 * from the *active* project's stores — that's the bug that used to
 * pin BeardGit's status under beardgit_gh_tests's key.
 *
 * Returns `null` for projects that have never been activated and
 * have no cache file. Callers that want fresh data for a non-active
 * project should use [`refreshProjectSnapshot`].
 */
export async function getSnapshotForHover(path: string): Promise<ProjectSnapshot | null> {
  return await loadProjectSnapshot(path);
}

/**
 * Force-refresh a project's snapshot via the `compute_project_snapshot`
 * Tauri command, which opens a temp repo handle at `path` on the Rust
 * side and reads its status without touching `AppState`. Updates both
 * the on-disk cache (server-side, via the command's persist step) and
 * the repo's in-memory `RepoState.snapshot` slice (client-side, here) so the
 * tab strip and tooltip both see fresh data.
 *
 * Used on tab mount for non-active projects to recover from any stale
 * cache (the previous broken fallback wrote active-project data under
 * inactive project keys; calling this once per non-active tab on
 * mount overwrites that).
 */
export async function refreshProjectSnapshot(path: string): Promise<ProjectSnapshot | null> {
  try {
    const snap = await api.computeProjectSnapshot(path);
    hydrateSnapshotCache(path, snap);
    return snap;
  } catch {
    return null;
  }
}

/**
 * Assemble a `graph_viewport_cache` slice from the current viewport
 * store, or return `null` when there's nothing worth persisting
 * (empty or absent viewport). Exported for unit tests so reconciliation
 * logic can assert shape parity without reaching into the store.
 */
export function buildGraphViewportCacheFromStores(
  headOid: string | null | undefined,
): NonNullable<ProjectSnapshot["graph_viewport_cache"]> | null {
  const vp = get(viewport);
  if (!vp || vp.nodes.length === 0) return null;
  const topOid = vp.nodes[0]?.oid ?? "";
  return {
    nodes: vp.nodes,
    total_count: vp.total_count,
    head_oid: headOid ?? "",
    top_oid: topOid,
    offset: get(graphOffset),
    cached_at: Date.now(),
  };
}

/** Build a snapshot from the current store state and save it. */
export async function saveCurrentSnapshot(projectPath: string): Promise<void> {
  const info = get(repoInfo);
  const statuses = get(fileStatuses);
  if (!info) return;

  // Use getStatusSummary for ahead/behind/stash data
  let ahead = 0, behind = 0, stash_count = 0, conflicted = 0;
  let staged = 0, unstaged = 0, untracked = 0;
  try {
    const s = await api.getStatusSummary();
    ahead = s.ahead;
    behind = s.behind;
    stash_count = s.stash_count;
    conflicted = s.conflicted;
    staged = s.staged;
    unstaged = s.unstaged;
    untracked = s.untracked;
  } catch { /* use defaults */ }

  const graph_viewport_cache = buildGraphViewportCacheFromStores(info.head_oid);

  // Per-project view memory: record which sidebar view THIS repo was left
  // on so a restart restores it too (the in-memory RepoState slice only
  // covers the current session).
  const repoSlice = getRepoState(projectPath);
  const activeView = repoSlice ? get(repoSlice.lastView) || null : null;

  const snapshot: ProjectSnapshot = {
    path: projectPath,
    head_branch: info.head_branch ?? null,
    ahead,
    behind,
    staged,
    unstaged,
    untracked,
    conflicted,
    stash_count,
    change_count: statuses.length,
    graph_viewport_cache,
    active_view: activeView,
  };

  // Mirror into memory first so subsequent tab switches hydrate instantly
  // without racing the save RTT.
  hydrateSnapshotCache(projectPath, snapshot);

  try {
    await api.saveProjectSnapshot(snapshot);
  } catch { /* non-critical */ }
}

File: crates/storage/src/project_cache.rs
//! Per-project snapshot cache for instant UI on project switch.
//!
//! Each project gets a small JSON file in `<config_dir>/project-cache/`
//! keyed by a hash of the project path. The snapshot stores the last-known
//! git state so the frontend can display badges, titlebar, and tooltip data
//! instantly without waiting for a full status computation.

use std::fs;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::StorageError;

/// Per-project cached git state for instant UI.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectSnapshot {
    /// Absolute path to the project's working directory.
    pub path: String,
    /// Current branch name (None if detached HEAD).
    pub head_branch: Option<String>,
    /// Commits ahead of upstream.
    pub ahead: usize,
    /// Commits behind upstream.
    pub behind: usize,
    /// Staged file count.
    pub staged: usize,
    /// Modified (unstaged) file count.
    pub unstaged: usize,
    /// Untracked file count.
    pub untracked: usize,
    /// Conflicted file count.
    pub conflicted: usize,
    /// Stash entry count.
    pub stash_count: usize,
    /// Total change count (staged + unstaged + untracked).
    pub change_count: usize,
    /// Persisted commit-graph viewport slice — populated on save so a
    /// cold start can paint the canvas synchronously. `#[serde(default)]`
    /// keeps older on-disk snapshots (pre-Phase-8) deserialising cleanly.
    #[serde(default)]
    pub graph_viewport_cache: Option<GraphViewportCache>,
    /// Last sidebar view the user browsed this project on (`"branches"`,
    /// `"changes"`, …). Restored per-project on tab activation so
    /// switching projects — and app restarts — land back on the view
    /// each repo was left on. `#[serde(default)]` keeps legacy snapshots
    /// deserialising cleanly; `None` = never recorded.
    #[serde(default)]
    pub active_view: Option<String>,
}

/// Persisted commit-graph viewport slice (see spec's cache shape).
///
/// Stored opaquely as `serde_json::Value` for `nodes` so the Rust side
/// doesn't need to mirror the `LayoutNode` struct one-to-one; the TS
/// layer owns the exact shape and drives validation. Size ≈ 60 KB.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct GraphViewportCache {
    /// Opaque `LayoutNode[]` JSON — Rust stores and forwards only.
    pub nodes: serde_json::Value,
    /// Total commit count at cache time.
    pub total_count: usize,
    /// HEAD OID at cache time; coarse staleness check.
    pub head_oid: String,
    /// First visible commit OID — primary scroll anchor for reconciliation.
    pub top_oid: String,
    /// Scroll offset captured at cache time.
    pub offset: usize,
    /// Epoch milliseconds when the cache was written; compared against
    /// the frontend's 7-day TTL on load.
    pub cached_at: i64,
}

/// Compute the cache filename for a project path.
///
/// Uses SHA-256 (truncated to 16 hex chars) rather than `DefaultHasher`,
/// whose output is not stable across Rust releases — a toolchain bump would
/// silently orphan every cached snapshot. A stable hash keeps a project's
/// cache file addressable across upgrades. (The one-time upgrade from the old
/// `DefaultHasher` scheme orphans existing snapshots once; they recompute.)
fn cache_filename(project_path: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(project_path.as_bytes());
    let digest = hasher.finalize();
    let mut out = String::with_capacity(16);
    for byte in digest.iter().take(8) {
        out.push_str(&format!("{byte:02x}"));
    }
    out
}

/// Return the cache directory path, creating it if needed.
fn cache_dir(config_dir: &Path) -> Result<PathBuf, StorageError> {
    let dir = config_dir.join("project-cache");
    if !dir.exists() {
        fs::create_dir_all(&dir)?;
    }
    Ok(dir)
}

/// Load a project snapshot from the cache. Returns `None` if not found.
pub fn load_snapshot(
    config_dir: &Path,
    project_path: &str,
) -> Result<Option<ProjectSnapshot>, StorageError> {
    let dir = cache_dir(config_dir)?;
    let file = dir.join(format!("{}.json", cache_filename(project_path)));
    if !file.exists() {
        return Ok(None);
    }
    let content = fs::read_to_string(&file)?;
    let snapshot: ProjectSnapshot = serde_json::from_str(&content)?;
    Ok(Some(snapshot))
}

/// Save a project snapshot to the cache.
pub fn save_snapshot(config_dir: &Path, snapshot: &ProjectSnapshot) -> Result<(), StorageError> {
    let dir = cache_dir(config_dir)?;
    let file = dir.join(format!("{}.json", cache_filename(&snapshot.path)));
    let content = serde_json::to_string_pretty(snapshot)?;
    fs::write(&file, content)?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cache_filename_deterministic() {
        let a = cache_filename("/Users/test/project");
        let b = cache_filename("/Users/test/project");
        assert_eq!(a, b);
        assert_eq!(a.len(), 16);
    }

    /// Pin the hash to a fixed value so a future toolchain bump (or an
    /// accidental swap back to a non-stable hasher) is caught. This is the
    /// first 8 bytes of `SHA-256("/Users/test/project")`.
    #[test]
    fn test_cache_filename_is_stable_sha256() {
        assert_eq!(cache_filename("/Users/test/project"), "0654434d556baf69");
    }

    #[test]
    fn test_cache_filename_different_paths() {
        let a = cache_filename("/Users/test/project-a");
        let b = cache_filename("/Users/test/project-b");
        assert_ne!(a, b);
    }

    #[test]
    fn test_save_and_load_snapshot() {
        let tmp = tempfile::tempdir().unwrap();
        let snapshot = ProjectSnapshot {
            path: "/Users/test/project".to_string(),
            head_branch: Some("main".to_string()),
            ahead: 2,
            behind: 0,
            staged: 1,
            unstaged: 3,
            untracked: 5,
            conflicted: 0,
            stash_count: 1,
            change_count: 9,
            graph_viewport_cache: None,
            active_view: None,
        };
        save_snapshot(tmp.path(), &snapshot).unwrap();
        let loaded = load_snapshot(tmp.path(), "/Users/test/project").unwrap();
        assert!(loaded.is_some());
        let loaded = loaded.unwrap();
        assert_eq!(loaded.path, "/Users/test/project");
        assert_eq!(loaded.ahead, 2);
        assert_eq!(loaded.change_count, 9);
        assert_eq!(loaded.head_branch, Some("main".to_string()));
        assert!(loaded.graph_viewport_cache.is_none());
    }

    #[test]
    fn test_load_missing_returns_none() {
        let tmp = tempfile::tempdir().unwrap();
        let loaded = load_snapshot(tmp.path(), "/nonexistent").unwrap();
        assert!(loaded.is_none());
    }

    #[test]
    fn test_overwrite_snapshot() {
        let tmp = tempfile::tempdir().unwrap();
        let snapshot1 = ProjectSnapshot {
            path: "/Users/test/project".to_string(),
            head_branch: Some("main".to_string()),
            ahead: 1,
            behind: 0,
            staged: 0,
            unstaged: 0,
            untracked: 0,
            conflicted: 0,
            stash_count: 0,
            change_count: 0,
            graph_viewport_cache: None,
            active_view: None,
        };
        save_snapshot(tmp.path(), &snapshot1).unwrap();

        let snapshot2 = ProjectSnapshot {
            path: "/Users/test/project".to_string(),
            head_branch: Some("feature".to_string()),
            ahead: 5,
            behind: 2,
            staged: 3,
            unstaged: 1,
            untracked: 0,
            conflicted: 0,
            stash_count: 0,
            change_count: 4,
            graph_viewport_cache: None,
            active_view: None,
        };
        save_snapshot(tmp.path(), &snapshot2).unwrap();

        let loaded = load_snapshot(tmp.path(), "/Users/test/project")
            .unwrap()
            .unwrap();
        assert_eq!(loaded.head_branch, Some("feature".to_string()));
        assert_eq!(loaded.ahead, 5);
    }

    /// Pre-Phase-8 snapshots on disk lack the `graph_viewport_cache`
    /// field. The `#[serde(default)]` attribute must let them
    /// deserialise without error so existing users don't lose their
    /// cached state on upgrade.
    #[test]
    fn test_legacy_snapshot_without_viewport_field_deserialises() {
        let legacy_json = r#"{
            "path": "/Users/test/legacy",
            "head_branch": "main",
            "ahead": 0,
            "behind": 0,
            "staged": 0,
            "unstaged": 0,
            "untracked": 0,
            "conflicted": 0,
            "stash_count": 0,
            "change_count": 0
        }"#;
        let snap: ProjectSnapshot = serde_json::from_str(legacy_json).unwrap();
        assert_eq!(snap.path, "/Users/test/legacy");
        assert!(snap.graph_viewport_cache.is_none());
    }

    /// Round-trip a snapshot with a populated viewport slice to verify
    /// the new field persists through save → load.
    #[test]
    fn test_snapshot_with_viewport_cache_roundtrips() {
        let tmp = tempfile::tempdir().unwrap();
        let snap = ProjectSnapshot {
            path: "/Users/test/vp".to_string(),
            head_branch: Some("main".to_string()),
            ahead: 0,
            behind: 0,
            staged: 0,
            unstaged: 0,
            untracked: 0,
            conflicted: 0,
            stash_count: 0,
            change_count: 0,
            graph_viewport_cache: Some(GraphViewportCache {
                nodes: serde_json::json!([{ "oid": "abc123" }]),
                total_count: 42,
                head_oid: "abc123".to_string(),
                top_oid: "abc123".to_string(),
                offset: 7,
                cached_at: 1_700_000_000_000,
            }),
            active_view: Some("branches".to_string()),
        };
        save_snapshot(tmp.path(), &snap).unwrap();
        let loaded = load_snapshot(tmp.path(), "/Users/test/vp")
            .unwrap()
            .unwrap();
        assert_eq!(loaded.active_view.as_deref(), Some("branches"));
        let cache = loaded.graph_viewport_cache.expect("cache should persist");
        assert_eq!(cache.total_count, 42);
        assert_eq!(cache.top_oid, "abc123");
        assert_eq!(cache.offset, 7);
        assert_eq!(cache.cached_at, 1_700_000_000_000);
    }
}
